### PR TITLE
ENH: Update deterministic tractography

### PR DESCRIPTION
**Pull request by kaitj**
Related to #105.This adds an exercise to change stopping criteria to `BinaryStoppingCriterion` for tracking and a quick visualization exercise to change the color of streamlines. I think this is the simplest alternative for the stopping criteria suggested in #105. ACT may be good to add in the future as a challenging exercise or more advanced lesson.

~~Will mark ready for review once I figure out the rotations for the cameras so I can add a figure to the solution output.~~

### CODE REVIEW

**Comment by kaitj** [comment]
PR should be ready to review now. I am open to adding an exercise for the probabilistic tractography if we can think of one, but given what has been taught in the current episodes, this may a good enough state for now.

File: _episodes/deterministic_tractography.md [hunk 1]
```diff
@@ -224,4 +224,99 @@ plt.show()
 
 ![EuDX Determinsitic Tractography]({{ relative_root_path }}/fig/deterministic_tractography/tractogram_deterministic_EuDX.png){:class="img-responsive"}
 
+> ## Exercise 1
+> 
+> In this episode, we used applied a threshold stopping criteria
```
Comment (jhlegarreta) [inline _episodes/deterministic_tractography.md:229]: `we used applied`: one of the two verbs is not necessary.

File: _episodes/deterministic_tractography.md [hunk 2]
```diff
@@ -224,4 +224,99 @@ plt.show()
...
+> > dti_fit = dti_model.fit(dwi_data, mask=dwi_mask)  
+> > fa_img = dti_fit.fa
+> > evecs_img = dti_fit.evecs
+> > 
+> > sphere = get_sphere('symmetric362')
+> > peak_indices = peaks_from_model(model=dti_model, data=dwi_data, sphere=sphere, relative_peak_threshold=.2, min_separation_angle=25, mask=dwi_mask, npeaks=2)
+> > 
+> > # Create a binary seed mask
+> > seed_mask = fa_img.copy()
+> > seed_mask[seed_mask>=0.2] = 1
```
Comment (jhlegarreta) [inline _episodes/deterministic_tractography.md:285]: Operators require whitespaces around according to PEP8.

File: _episodes/deterministic_tractography.md [hunk 3]
```diff
@@ -224,4 +224,99 @@ plt.show()
...
+> > seed_mask[seed_mask<0.2] = 0
```
Comment (jhlegarreta) [inline _episodes/deterministic_tractography.md:286]: Same comment as above.

**Review by jhlegarreta — COMMENTED** [review]
Looks a fair exercise taking into account our time constraints.

As for the visualization, I was more thinking in #105 to apply the FA scalar map rather than applying the same color and just changing the opacity to a value of 0.2 (which is IMHO misleading as it is the one used for the FA). But as I said, given our constraints, it is fine the way is set up.

IMHO the notebooks should be modified to include the exercise within this PR, otherwise we'll again run the risk of having out-of-sync Markdown vs. notebooks.

As for the prob tracking episode, I'm fine if we do not add an exercise; I've put a note in the issue with other relatively easy candidates.

**Comment by jhlegarreta** [comment]
Please consider merging #123, #124 and #125 before, and consider using the `get_fdata()` call in [here](https://github.com/carpentries-incubator/SDC-BIDS-dMRI/pull/120/files#diff-df2aeeda2a19bf81a529c7116412c493977777c2c212578157a4a9bdb9ccf89dR271) and in any other eventual call (e.g. when changing the notebooks as requested above).

File: _episodes/deterministic_tractography.md [hunk 4]
```diff
@@ -224,4 +224,99 @@ plt.show()
...
+> > bvec_fname = dwi_layout.get(subject=subj, extension='eddy_rotated_bvecs', return_type='file')[0]
+> > bval_fname = gradient_layout.get(subject=subj, suffix='dwi', extension='bval', return_type='file')[0]
+> > 
+> > dwi_img = nib.load(dwi_fname)
+> > affine = dwi_img.affine
+> > 
+> > bvals, bvecs = read_bvals_bvecs(bval_fname, bvec_fname)
+> > gtab = gradient_table(bvals, bvecs)
+> > 
+> > dwi_data = dwi_img.get_data()
```
Comment (jhlegarreta) [inline _episodes/deterministic_tractography.md:271]: Cross-referencing #125.

**Comment by kaitj** [comment]
Going to change this to draft now - planning on sitting down one of the nights next week to just solely update this episode with figures, sync content between markdown and notebook(s) and run a linter through the notebook.

**Comment by jhlegarreta** [comment]
3a3069d is not necessary:
https://carpentries-incubator.github.io/SDC-BIDS-dMRI/local_tractography/index.html

The people behind Markdown made life a little bit easer.

**Comment by kaitj** [comment]
> [3a3069d](https://github.com/carpentries-incubator/SDC-BIDS-dMRI/commit/3a3069d2e5c97e9fd329a9309b175ee6515077e8) is not necessary:
> https://carpentries-incubator.github.io/SDC-BIDS-dMRI/local_tractography/index.html
> 
> The people behind Markdown made life a little bit easer.

Ahh, thats good to know. Will revert the commit.

**Comment by kaitj** [comment]
PR should be ready to review now. Markdown and notebooks should be synced - I removed the latex call as it doesn't render properly in the markdown - we can add it back in later on as an image if needed. 

Exercise now asks the learner to color each point of the streamline with the FA map with addition of having to transform the streamlines from world to native space. I believe these different spaces were covered in the IntroMRI so I didn't include any additional detail aside from the necessary command.

Also used this PR to update all of the deterministic tractography figures (see #131).

File: _episodes/deterministic_tractography.md [hunk 5]
```diff
@@ -228,4 +228,105 @@ plt.show()
...
+streamlines = Streamlines(streamlines_generator)
+> > ~~~
+> > {: .language-python}
+> {: .solution}
+> 
+> ## Exercise 2
+> 
+>Visualize the tractogram! As an additional challenge, set the
+> color of the streamlines to display the values of the
+FA map
```
Comment (jhlegarreta) [inline _episodes/deterministic_tractography.md:308]: Undesired linebreak. May even break the `challenge` environment.

File: _episodes/deterministic_tractography.md [hunk 6]
```diff
@@ -228,4 +228,105 @@ plt.show()
...
+> > seed_mask[seed_mask < 0.2] = 0
+> > 
+> > seeds = utils.seeds_from_mask(seed_mask, affine=affine, density=1)
+> > 
+> > # Set stopping criteria
+> > stopping_criterion = BinaryStoppingCriterion(seed_mask==1)
+> > 
+> > # Perform tracking
+> > streamlines_generator = LocalTracking(peak_indices, stopping_criterion, seeds, affine=affine, step_size=.5)
+streamlines = Streamlines(streamlines_generator)
```
Comment (jhlegarreta) [inline _episodes/deterministic_tractography.md:299]: Not sure if this linebreak will break the `challenge` environment.

File: _episodes/deterministic_tractography.md [hunk 7]
```diff
@@ -228,4 +228,105 @@ plt.show()
...
+> {: .solution}
+> 
+> ## Exercise 2
+> 
+>Visualize the tractogram! As an additional challenge, set the
+> color of the streamlines to display the values of the
+FA map 
+> and change the opacity to `0.05`. You may need to transform 
+> the streamlines from world coordinates to the subject's native space 
+>using `transform_streamlines` from `dipy.tracking.streamline`.
```
Comment (jhlegarreta) [inline _episodes/deterministic_tractography.md:311]: Similar comments as above.

**Review by jhlegarreta — COMMENTED** [review]
I haven't checked the exercises, but fine for merging if you wish.

LaTeX code not rendering might eventually be fixed thanks to this https://github.com/carpentries/carpentries-theme/pull/13, but fine removing until then.

I think the picture added at the end corresponds just to the first exercise, and the picture corresponding to the second one is missing. I would add the picture corresponding to each exercise right after the exercise and before the next exercise if that is possible.

File: _episodes/deterministic_tractography.md (hunk 5 above)
Comment (jhlegarreta) [reply _episodes/deterministic_tractography.md]: Fixed in b886a5a.

File: _episodes/deterministic_tractography.md (hunk 6 above)
Comment (jhlegarreta) [reply _episodes/deterministic_tractography.md]: Fixed in b886a5a.

File: _episodes/deterministic_tractography.md (hunk 7 above)
Comment (jhlegarreta) [reply _episodes/deterministic_tractography.md]: Fixed in b886a5a.

**Comment by jhlegarreta** [comment]
> As for the image, I didn't include a screenshot for exercise 1 given we did not ask for visualization unless you are talking about something different?

Can be added in a separate PR, but if we do not visualize we do not have any visual clue of how the tracking performed. Learners might be told to just reuse/copy the visualization lines in the episode to show the result of the tracking.

**Comment by kaitj** [comment]
Added tract visualization and fixed spelling mistake to exercise 1 in 18d790f

**Review by jhlegarreta — APPROVED** [review]

**Pull request merged**